File: app/src/main/java/com/carbonplayer/model/network/entity/DownloadRequest.java
package com.carbonplayer.model.network.entity;

import com.carbonplayer.model.entity.SongID;
import com.carbonplayer.model.entity.enums.StorageType;
import com.carbonplayer.model.entity.enums.StreamQuality;


public class DownloadRequest {

    public static int PRIORITY_AUTOCACHE = 200;
    public static int PRIORITY_KEEPON = 100;
    public static int PRIORITY_PREFETCH1 = 1;
    public static int PRIORITY_PREFETCH2 = 2;
    public static int PRIORITY_PREFETCH3 = 3;
    public static int PRIORITY_PREFETCH4 = 4;
    public static int PRIORITY_STREAM = 0;

    private SongID id;

    private final StreamQuality existingQuality;
    private final StreamQuality requestedQuality;
    private final long seekMillis;
    private final String trackTitle;
    private FileLocation fileLocation;

    private State state;

    enum State {
        NOT_STARTED,
        DOWNLOADING,
        COMPLETED,
        FAILED,
        CANCELED
    }

    public DownloadRequest(SongID id, String trackTitle, int priority, long seekMillis, FileLocation fileLocation, boolean explicit, StreamQuality requestedQuality, StreamQuality existingQuality) {
        if (trackTitle == null) {
            throw new IllegalArgumentException("The track title is required");
        } else if (seekMillis < 0) {
            throw new IllegalArgumentException("Negative seek time: " + seekMillis);
        } else if (requestedQuality == StreamQuality.UNDEFINED) {
            throw new IllegalArgumentException("Requested quality cannot be unknown");
        } else if (existingQuality == StreamQuality.UNDEFINED || requestedQuality.compareTo(existingQuality) > 0) {
            this.id = id;
            this.trackTitle = trackTitle;
            this.seekMillis = seekMillis;
            this.requestedQuality = requestedQuality;
            this.existingQuality = existingQuality;
            this.fileLocation = fileLocation;
        } else {
            throw new IllegalArgumentException("If existing quality is known, requested quality must exceed it: existing=" + existingQuality + " requested=" + requestedQuality);
        }
    }

    public SongID getId() {
        return this.id;
    }

    public String getTrackTitle() {
        return this.trackTitle;
    }

    public long getSeekMillis() {
        return this.seekMillis;
    }

    public FileLocation getFileLocation(){
        return fileLocation;
    }

    public boolean equals(Object obj) {
        if (!(obj instanceof DownloadRequest)) {
            return false;
        }
        DownloadRequest downloadRequest = (DownloadRequest) obj;
        if (this.id == downloadRequest.id && this.seekMillis == downloadRequest.getSeekMillis()) {
            return true;
        }
        return false;
    }

    public String toString() {
        return "DownloadRequest{id='" + this.id + '\'' + ", mSourceAccount=" + '\'' + ", trackTitle='" + this.trackTitle + '\'' + ", seekMillis=" + this.seekMillis + "} ";
    }

    public State getState(){
        return state;
    }

    public void setState(State state) {this.state = state;}

    protected int getMinPriority() {
        return PRIORITY_AUTOCACHE;
    }

    protected int getMaxPriority() {
        return PRIORITY_STREAM;
    }

    public StreamQuality getRequestedQuality() {
        return this.requestedQuality;
    }

    public StreamQuality getExistingQuality() {
        return this.existingQuality;
    }
}

File: app/src/main/java/com/carbonplayer/model/network/entity/DownloadRequest.kt
package com.carbonplayer.model.network.entity

import com.carbonplayer.model.entity.SongID
import com.carbonplayer.model.entity.enums.StreamQuality


data class DownloadRequest(
        var id: SongID? = null,
        var trackTitle: String,
        val priority: Int,
        val seekMillis: Long,
        var fileLocation: FileLocation,
        val explicit: Boolean,
        val requestedQuality: StreamQuality,
        val existingQuality: StreamQuality
) {

    var state: State? = null

    private val minPriority: Int get() = PRIORITY_AUTOCACHE
    private val maxPriority: Int get() = PRIORITY_STREAM

    enum class State {
        NOT_STARTED,
        DOWNLOADING,
        COMPLETED,
        FAILED,
        CANCELED
    }

    init {
        if (seekMillis < 0) {
            throw IllegalArgumentException("Negative seek time: " + seekMillis)
        } else if (requestedQuality == StreamQuality.UNDEFINED) {
            throw IllegalArgumentException("Requested quality cannot be unknown")
        } else if (existingQuality != StreamQuality.UNDEFINED || requestedQuality <= existingQuality) {
            throw IllegalArgumentException(
                    "Existing quality ${existingQuality.name} must exceed requested quality ${requestedQuality.name}"
            )
        }
    }

    override fun equals(other: Any?): Boolean {
        if (other !is DownloadRequest) return false
        return this.id === other.id && this.seekMillis == other.seekMillis
    }

    companion object {
        var PRIORITY_AUTOCACHE = 200
        var PRIORITY_KEEPON = 100
        var PRIORITY_PREFETCH1 = 1
        var PRIORITY_PREFETCH2 = 2
        var PRIORITY_PREFETCH3 = 3
        var PRIORITY_PREFETCH4 = 4
        var PRIORITY_STREAM = 0
    }
}
